pkg/config/structure/unmarshal.go

```diff
@@ -0,0 +1,368 @@
+	switch n.val.Kind() {
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return int(n.val.Int()), nil
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		return int(n.val.Uint()), nil
+	}
+	return 0, conversionError(n.val, "int")
+}
+
+// GetString returns the scalar as a string, or an error otherwise
+func (n *leafimpl) GetString() (string, error) {
+	if n.val.Kind() == reflect.String {
+		return n.val.String(), nil
+	}
+	return "", conversionError(n.val, "string")
+}
+
+func copyStruct(target reflect.Value, source node) error {
+	targetType := target.Type()
+	for i := 0; i < targetType.NumField(); i++ {
+		f := targetType.Field(i)
+		ch, _ := utf8.DecodeRuneInString(f.Name)
+		if unicode.IsLower(ch) {
+			continue
+		}
+		key := f.Name
+		if replace := f.Tag.Get("yaml"); replace != "" {
+			key = replace
+		} else if replace := f.Tag.Get("json"); replace != "" {
+			key = replace
+		} else if replace := f.Tag.Get("mapstructure"); replace != "" {
+			key = replace
+		}
+		child := source.GetChild(key)
+		if child == nil {
+			continue
+		}
+		err := copyAny(target.FieldByName(f.Name), child)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func copyMap(target reflect.Value, source node) error {
+	// TODO: Should handle maps with more complex types
+	ktype := reflect.TypeOf("")
+	vtype := reflect.TypeOf("")
+	mtype := reflect.MapOf(ktype, vtype)
+	results := reflect.MakeMap(mtype)
+
+	mapKeys := source.ChildrenKeys()
+	for _, mkey := range mapKeys {
+		child := source.GetChild(mkey)
+		if child == nil {
+			continue
+		}
+		scalar := child.AsScalar()
+		if scalar != nil {
+			mval, _ := scalar.GetString()
+			results.SetMapIndex(reflect.ValueOf(mkey), reflect.ValueOf(mval))
+		}
+	}
+	target.Set(results)
+	return nil
+}
+
+func copyScalar(target reflect.Value, source leaf) error {
+	if source == nil {
+		return fmt.Errorf("source value is not a scalar")
+	}
+	switch target.Kind() {
+	case reflect.Bool:
+		v, err := source.GetBool()
+		if err != nil {
+			return err
+		}
+		target.SetBool(v)
+		return nil
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		v, err := source.GetInt()
+		if err != nil {
+			return err
+		}
+		target.SetInt(int64(v))
+		return nil
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		v, err := source.GetInt()
+		if err != nil {
+			return err
+		}
+		target.SetUint(uint64(v))
+		return nil
+	case reflect.String:
+		v, err := source.GetString()
+		if err != nil {
+			return err
+		}
+		target.SetString(v)
+		return nil
+	}
+	return fmt.Errorf("unsupported scalar type %v", target.Kind())
+}
+
+func copyList(target reflect.Value, source nodeList) error {
+	if source == nil {
+		return fmt.Errorf("source value is not a list")
+	}
+	elemType := target.Type()
+	elemType = elemType.Elem()
+	results := reflect.MakeSlice(reflect.SliceOf(elemType), source.Size(), source.Size())
```

Comment:
We are accessing `souorce.Size()` three times, should hosting the size value on a variable?

Reply:
Sure, done, now `numElems` is used for this value.